e2e/tests-dfx/ledger.bash

```diff
@@ -117,6 +137,18 @@ tc_to_num() {
     balance_now=$(tc_to_num "$(dfx wallet balance)")
     
     (( balance_now - balance > 600000000000000 ))
+    
+    # Transaction Deduplication
+    t=$(current_time_nanoseconds)
+
+    assert_command dfx ledger top-up "$wallet" --icp 5 --created-at-time "$t"
+    assert_match "Canister was topped up with"
+
+    assert_command dfx ledger top-up "$wallet" --icp 5 --created-at-time $((t+1))
+    assert_match "Canister was topped up with"
+
+    run dfx ledger top-up "$wallet" --icp 5 --created-at-time "$t"
```

Comment:
what is `run`?

```suggestion
    assert_command dfx ledger top-up "$wallet" --icp 5 --created-at-time "$t"
```

Reply:
`run` is the way you’d run the command in up-to-date `bats`. With, e.g. bats 1.9, instead of assert_match, you’d split it two lines
```bash
run CMD
assert_success  # imported from bats-assert
```
I did try to see how difficult would it be to move us over to new version of bats here: https://github.com/smallstepman/sdk/pull/46 but the work was left unfinished, though I am planning to pick it up again at some point

Reply:
> I thought the command would fail because it is a duplicate transaction, do you know why it succeeds anyway?

From the perspective of the `dfx ledger top-up` command on line 150, it (the transfer portion, at least) is a duplicate of the `dfx ledger top-up` command on line 144.  They share the same `created_at_time`, `memo`, `amount`, and so forth.

Also from the perspective of the second invocation of `dfx ledger top-up` with the same parameters, we don't know that the previous call succeeded.  A plausible reason for running `dfx ledger top-up` with the exact same parameters would be that we are unsure whether the first command succeeded, or that part of it succeeded (`transfer_cmc`), and part of it failed (`notify_top_up`).

So in the context of the second call to `dfx ledger top-up` with the exact same parameters, dfx sees the `TxDuplicate` result, continues and reports success.

Reply:
@smallstepman Either there or a stackoverflow post. The `assert_command_fail` from the utils file is a better fit for what I was trying to do.

Reply:
> So in the context of the second call to dfx ledger top-up with the exact same parameters, dfx sees the TxDuplicate result, continues and reports success.

@ericswanson-dfinity Why is the notify part of the command succeeding? 
https://github.com/dfinity/ic/blob/3323f0fea730cef34233f6da48bbf043af5f0da4/rs/nns/cmc/src/main.rs#L1026

Reply:
> So in the context of the second call to dfx ledger top-up with the exact same parameters, dfx sees the TxDuplicate result, continues and reports success.

@ericswanson-dfinity I was wondering why the notify part succeeds even if it's the second time calling notify with the same block height but I see [here](https://github.com/dfinity/ic/blob/3323f0fea730cef34233f6da48bbf043af5f0da4/rs/nns/cmc/src/main.rs#L1026) that the cmc returns the cached result of the first notification of that block height which is a success. Thanks.